src/input/src/lib.rs

```diff
@@ -114,7 +119,167 @@ pub enum Input {
     /// Window gained or lost cursor.
     Cursor(bool),
     /// Window closed.
-    Close
+    Close(CloseArgs),
+    /// Render graphics.
+    Render(RenderArgs),
+    /// After rendering and swapping buffers.
+    AfterRender(AfterRenderArgs),
+    /// Update the state of the application.
+    Update(UpdateArgs),
+    /// Do background tasks that can be done incrementally.
+    Idle(IdleArgs),
+    /// Custom event.
+    ///
+    /// When comparing two custom events for equality,
+    /// they always return `false`.
+    Custom(EventId, Arc<Any>),
```

Comment:
Thanks! I'll make a PR. Also updated https://github.com/PistonDevelopers/piston/issues/1150 to make it clearer:

> This will make it easier to add custom events, without needing a separate implementation of `GenericEvent`.